### PR TITLE
DM-43209: Add type annotations

### PR DESCRIPTION
**Pull request by kkeijuro**
Add type annotations to summit_extras

### CODE REVIEW

File: python/lsst/summit/extras/fastStarTrackerAnalysis.py [hunk 1]
```diff
@@ -359,7 +370,7 @@ def checkResultConsistency(results, maxAllowableShift=5, silent=False):
     Parameters
     ----------
     results : `dict` of `list` of
-              `lsst.summit.extras.fastStarTrackerAnalysis.Source`
+              `lsst.summit.utils.astrometry.fastStarTrackerAnalysis.Source`
```
Comment (mfisherlevine) [inline python/lsst/summit/extras/fastStarTrackerAnalysis.py:373]: Rebase error.

File: python/lsst/summit/extras/fastStarTrackerAnalysis.py [hunk 2]
```diff
@@ -463,7 +478,7 @@ def plotSourceMovement(results, sourceIndex=0, allowInconsistent=False):
...
```
Comment (mfisherlevine) [inline python/lsst/summit/extras/fastStarTrackerAnalysis.py:481]: Rebase error.

File: python/lsst/summit/extras/fastStarTrackerAnalysis.py [hunk 3]
```diff
@@ -575,16 +590,19 @@ def plotSourceMovement(results, sourceIndex=0, allowInconsistent=False):
...
 
     Parameters
     ----------
     parentFilename : `str`
         The full path to the parent (.tif) file.
     sources : `list` of
-              `lsst.summit.extras.fastStarTrackerAnalysis.Source` or
-              `lsst.summit.extras.fastStarTrackerAnalysis.Source`
+              `lsst.summit.utils.astrometry.fastStarTrackerAnalysis.Source` or
+              `lsst.summit.utils.astrometry.fastStarTrackerAnalysis.Source`
```
Comment (mfisherlevine) [inline python/lsst/summit/extras/fastStarTrackerAnalysis.py:605]: Rebase error.

File: python/lsst/summit/extras/fastStarTrackerAnalysis.py [hunk 4]
```diff
@@ -614,12 +632,12 @@ def plotSourcesOnImage(parentFilename, sources):
...
 
 
-def plotSource(source):
+def plotSource(source: summitExtras.fastStarTrackerAnalysis.Source) -> None:
     """Plot a single source.
 
     Parameters
     ----------
-    source : `lsst.summit.extras.fastStarTrackerAnalysis.Source`
+    source : `lsst.summit.utils.astrometry.fastStarTrackerAnalysis.Source`
```
Comment (mfisherlevine) [inline python/lsst/summit/extras/fastStarTrackerAnalysis.py:640]: Rebase error.

File: python/lsst/summit/extras/focusAnalysis.py [hunk 5]
```diff
@@ -90,18 +93,18 @@ class SpectralFocusAnalyzer:
...
 
         self.spectrumHalfWidth = 100
         self.spectrumBoxLength = 20
-        self._spectrumBoxOffsets = [882, 1170, 1467]
+        self._spectrumBoxOffsets = [882.0, 1170.0, 1467.0]
         self._setColors(len(self._spectrumBoxOffsets))
 
-    def setSpectrumBoxOffsets(self, offsets):
+    def setSpectrumBoxOffsets(self, offsets: List[float]) -> None:
         """Set the current spectrum slice offsets.
```
Comment (mfisherlevine) [inline python/lsst/summit/extras/focusAnalysis.py:108]: These actually are integers, and need to be typed as such, as they're used for de-referencing arrays (or the equivalent, actually used in Box2I construction, but it's the same thing.)

File: python/lsst/summit/extras/focusAnalysis.py [hunk 6]
```diff
@@ -113,7 +116,7 @@ def setSpectrumBoxOffsets(self, offsets):
         self._spectrumBoxOffsets = offsets
         self._setColors(len(offsets))
 
-    def getSpectrumBoxOffsets(self):
+    def getSpectrumBoxOffsets(self) -> List[float]:
```
Comment (mfisherlevine) [inline python/lsst/summit/extras/focusAnalysis.py:119]: int

File: python/lsst/summit/extras/focusAnalysis.py [hunk 7]
```diff
@@ -124,10 +127,10 @@ def getSpectrumBoxOffsets(self):
         """
         return self._spectrumBoxOffsets
 
-    def _setColors(self, nPoints):
+    def _setColors(self, nPoints: int) -> None:
         self.COLORS = cm.rainbow(np.linspace(0, 1, nPoints))
 
-    def _getBboxes(self, centroid):
+    def _getBboxes(self, centroid: "List[float]") -> "geom.Box2I":
```
Comment (mfisherlevine) [inline python/lsst/summit/extras/focusAnalysis.py:133]: Remove strings from typing.

File: python/lsst/summit/extras/focusAnalysis.py [hunk 8]
```diff
@@ -150,11 +153,18 @@ def _bboxToMplRectangle(self, bbox, colorNum):
         return rectangle
 
     @staticmethod
-    def gauss(x, *pars):
+    def gauss(x: float, *pars: float) -> float:
```
Comment (mfisherlevine) [inline python/lsst/summit/extras/focusAnalysis.py:156]: `x` is indeed a float. Is that how you annotate catching a list of other parameters though?

File: python/lsst/summit/extras/focusAnalysis.py [hunk 9]
```diff
@@ -150,11 +153,18 @@ def _bboxToMplRectangle(self, bbox, colorNum):
...
         amp, mean, sigma = pars
         return amp * np.exp(-((x - mean) ** 2) / (2.0 * sigma**2))
 
-    def run(self, dayObs, seqNums, doDisplay=False, hideFit=False, hexapodZeroPoint=0):
+    def run(
+        self,
+        dayObs: int,
+        seqNums: "List[int]",
```
Comment (mfisherlevine) [inline python/lsst/summit/extras/focusAnalysis.py:163]: No strings for types.

File: python/lsst/summit/extras/focusAnalysis.py [hunk 10]
```diff
@@ -406,25 +416,25 @@ class NonSpectralFocusAnalyzer:
...
     """
 
-    def __init__(self, embargo=False):
+    def __init__(self, embargo: bool = False):
         self.butler = makeDefaultLatissButler(embargo=embargo)
         self._bestEffort = BestEffortIsr(embargo=embargo)
 
     @staticmethod
-    def gauss(x, *pars):
+    def gauss(x: float, *pars: "float"):
```
Comment (mfisherlevine) [inline python/lsst/summit/extras/focusAnalysis.py:424]: Again, no strings, and also, is `float` actually correct for `*pars`?

File: python/lsst/summit/extras/focusAnalysis.py [hunk 11]
```diff
@@ -406,25 +416,25 @@ class NonSpectralFocusAnalyzer:
...
-        seqNums,
+        dayObs: int,
+        seqNums: List[int],
         *,
-        manualCentroid=None,
-        doCheckDispersed=True,
-        doDisplay=False,
-        doForceCoM=False,
-    ):
+        manualCentroid: Tuple[float] | None = None,
```
Comment (mfisherlevine) [inline python/lsst/summit/extras/focusAnalysis.py:433]: I think we can be more specific and make this a two-ple of floats, _i.e._ `(float, float)`, however that should be expressed, as it'll always be of length two.

File: python/lsst/summit/extras/focusAnalysis.py [hunk 12]
```diff
@@ -469,14 +479,14 @@ def run(
...
```
Comment (mfisherlevine) [inline python/lsst/summit/extras/focusAnalysis.py:485]: Same here.

File: python/lsst/summit/extras/imageSorter.py [hunk 13]
```diff
@@ -56,22 +57,22 @@ class ImageSorter:
...
     """
 
-    def __init__(self, fileList, outputFilename):
+    def __init__(self, fileList: List[str], outputFilename: str):
         self.fileList = fileList
         self.outputFilename = outputFilename
 
     @staticmethod
-    def _getDataIdFromFilename(filename):
+    def _getDataIdFromFilename(filename: str) -> "Tuple[str, int]":
```
Comment (mfisherlevine) [inline python/lsst/summit/extras/imageSorter.py:65]: No strings in types.

File: python/lsst/summit/extras/imageSorter.py [hunk 14]
```diff
@@ -168,7 +169,7 @@ def sortImages(self):
             else:
                 print("Unrecognised response - try again")
                 self.sortImages()
-                return  # don't run twice in this case!
+                return None  # don't run twice in this case!
```
Comment (mfisherlevine) [inline python/lsst/summit/extras/imageSorter.py:172]: Is this necessary?

File: python/lsst/summit/extras/logUtils.py [hunk 15]
```diff
@@ -74,7 +80,7 @@ def __init__(self, butler, taskName, collection):
         self.dataRefs = self._getDataRefs()
         self.logs = self._loadLogs(self.dataRefs)
 
-    def _getDataRefs(self):
+    def _getDataRefs(self) -> "List[dafButler.core.datasets.ref.DatasetRef]":
```
Comment (mfisherlevine) [inline python/lsst/summit/extras/logUtils.py:83]: No strings.

File: python/lsst/summit/extras/monitoring.py [hunk 16]
```diff
@@ -63,13 +66,13 @@ def __init__(self, fireflyDisplay, **kwargs):
...
 
-    def _getLatestImageDataIdAndExpId(self):
+    def _getLatestImageDataIdAndExpId(self) -> tuple:
         """Get the dataId and expId for the most recent image in the repo."""
         dataId = getMostRecentDataId(self.butler)
         expId = getExpIdFromDayObsSeqNum(self.butler, dataId)["exposure"]
         return dataId, expId
 
-    def _calcImageStats(self, exp):
+    def _calcImageStats(self, exp: "afwImage.Exposure") -> "List[str]":
```
Comment (mfisherlevine) [inline python/lsst/summit/extras/monitoring.py:75]: No strings for types.

File: python/lsst/summit/extras/slewTiming.py [hunk 17]
```diff
@@ -74,7 +80,13 @@
 ]
 
 
-def getMountPositionData(client, begin, end, prePadding=0, postPadding=0):
+def getMountPositionData(
+    client: EfdClient,
+    begin: astropy.time.Time,
+    end: astropy.time.Time,
+    prePadding: int = 0,
+    postPadding: int = 0,
```
Comment (mfisherlevine) [inline python/lsst/summit/extras/slewTiming.py:88]: Both can be floats too.

File: python/lsst/summit/extras/slewTiming.py [hunk 18]
```diff
@@ -119,7 +131,13 @@ def getMountPositionData(client, begin, end, prePadding=0, postPadding=0):
     return az, el, rot
 
 
-def getAxesInPosition(client, begin, end, prePadding, postPadding):
+def getAxesInPosition(
+    client: EfdClient,
+    begin: astropy.time.Time,
+    end: astropy.time.Time,
+    prePadding: int = 0,
+    postPadding: int = 0,
```
Comment (mfisherlevine) [inline python/lsst/summit/extras/slewTiming.py:139]: Can be floats.

File: python/lsst/summit/extras/slewTiming.py [hunk 19]
```diff
@@ -130,7 +148,13 @@ def getAxesInPosition(client, begin, end, prePadding, postPadding):
     )
 
 
-def plotExposureTiming(client, expRecords, plotHexapod=False, prePadding=1, postPadding=3):
+def plotExposureTiming(
+    client: EfdClient,
+    expRecords: dafButler.DimensionRecord,
```
Comment (mfisherlevine) [inline python/lsst/summit/extras/slewTiming.py:153]: It's an iterable of these.

File: python/lsst/summit/extras/slewTiming.py [hunk 20]
```diff
@@ -130,7 +148,13 @@ def getAxesInPosition(client, begin, end, prePadding, postPadding):
...
+    plotHexapod=False,
```
Comment (mfisherlevine) [inline python/lsst/summit/extras/slewTiming.py:154]: Missing the type `bool`.

File: python/lsst/summit/extras/slewTiming.py [hunk 21]
```diff
@@ -130,7 +148,13 @@ def getAxesInPosition(client, begin, end, prePadding, postPadding):
...
+    prePadding: int = 1,
+    postPadding: int = 3,
```
Comment (mfisherlevine) [inline python/lsst/summit/extras/slewTiming.py:156]: Can be floats.

File: python/lsst/summit/extras/animation.py [hunk 22]
```diff
@@ -26,11 +26,13 @@
 import shutil
 import subprocess
 import uuid
+from typing import Any, List, Tuple
```
Comment (timj) [inline python/lsst/summit/extras/animation.py:29]: Don't use `List` or `Tuple`. Use `list` and `tuple`.

**Comment by timj** [comment]
Two general comments:

1. Use `from __future__ import annotations` at the top of each file.
2. Don't use `Dict`, `Tuple` or `Set` etc. Use the standard python classes. You can update what you have with `ruff --select=UP --fix .`

File: python/lsst/summit/extras/animation.py [hunk 23]
```diff
@@ -213,7 +220,7 @@ def run(self):
         # stage files in temp dir with numbers prepended to filenames
         if not self.dataIdList:
             logger.warning("No files to animate - nothing to do")
-            return
+            return None
```
[the remark this answers is not in the document]
Comment (kkeijuro) [reply python/lsst/summit/extras/animation.py:219]: Mypy specifies as an error:

Return value expectedMypy[return-value](https://mypy.readthedocs.io/en/latest/_refs.html#code-return-value)

File: python/lsst/summit/extras/fastStarTrackerAnalysis.py [hunk 24]
```diff
@@ -43,27 +47,26 @@
 from lsst.utils.iteration import ensure_iterable
 
 __all__ = (
-    "getStreamingSequences",
     "getFlux",
     "getBackgroundLevel",
     "countOverThresholdPixels",
     "sortSourcesByFlux",
-    "Source",
-    "NanSource",
```
[the remark this answers is not in the document]
Comment (kkeijuro) [reply python/lsst/summit/extras/fastStarTrackerAnalysis.py:52]: I think so...

File: python/lsst/summit/extras/focusAnalysis.py (hunk 5 above)
Comment (kkeijuro) [reply python/lsst/summit/extras/focusAnalysis.py:108]: OK. In this case I was just following the docstring:

        offsets : `list` of `float`

File: python/lsst/summit/extras/focusAnalysis.py (hunk 8 above)
Comment (kkeijuro) [reply python/lsst/summit/extras/focusAnalysis.py:156]: Yes, is done that way for annotating a list of parameters.

File: python/lsst/summit/extras/imageSorter.py (hunk 14 above)
Comment (kkeijuro) [reply python/lsst/summit/extras/imageSorter.py:172]: Yes, to avoid mypy error

File: python/lsst/summit/extras/animation.py (hunk 22 above)
Comment (kkeijuro) [reply python/lsst/summit/extras/animation.py:29]: Nice! Thanks for the tips!

File: .pre-commit-config.yaml [hunk 25]
```diff
@@ -12,7 +12,7 @@ repos:
         additional_dependencies: [toml]
 
   - repo: https://github.com/psf/black
-    rev: 23.12.1
+    rev: 24.1.1
```
[the remark this answers is not in the document]
Comment (kkeijuro) [reply .pre-commit-config.yaml:15]: I just pick the last stable version

File: python/lsst/summit/extras/fastStarTrackerAnalysis.py [hunk 26]
```diff
@@ -116,7 +119,7 @@ def getStreamingSequences(dayObs):
     return data
 
 
-def getFlux(cutout, backgroundLevel=0):
+def getFlux(cutout: np.ndarray[int], backgroundLevel: int = 0) -> float:
```
[the remark this answers is not in the document]
Comment (mfisherlevine) [reply python/lsst/summit/extras/fastStarTrackerAnalysis.py]: `backgroundLevel` is still an `int` and it needs to be a `float`.

File: python/lsst/summit/extras/fastStarTrackerAnalysis.py [hunk 27]
```diff
@@ -345,7 +352,11 @@ def findFastStarTrackerImageSources(filename, boxSize, attachCutouts=True):
     return sortSourcesByFlux(sources)
 
 
-def checkResultConsistency(results, maxAllowableShift=5, silent=False):
+def checkResultConsistency(
+    results: Dict[int, List[summitExtras.fastStarTrackerAnalysis.Source]],
+    maxAllowableShift: int = 5,
```
[the remark this answers is not in the document]
Comment (mfisherlevine) [reply python/lsst/summit/extras/fastStarTrackerAnalysis.py:354]: Still needs fixing.

**Pull request merged**